### PR TITLE
async triggers

### PR DESCRIPTION
**Pull request by gagoar**
**Issue Reference**: #213 

## Description

all triggers are now async by default. 

## Motivation and Context
As a step to enable #346 and keep refactoring without changing functionality for #213  I've made all. the interfaces for the triggers be async. 

_note_:  Sadly I couldn't replace as much code as I wanted with the generated guards. but this will be solved when we are able to refactor the rules shape. 

for now, we will have to keep the function `isValidRawRule` around given that it is more expressive than the ones autogenerated. 

## How Has This Been Tested?
unit tests. 

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -1,21 +1,21 @@
-process.env.TZ = "UTC";
+process.env.TZ = 'UTC';
 
 module.exports = {
   globals: {
-    "ts-jest": {
+    'ts-jest': {
```
Comment (gagoar) [inline jest.config.js:5]: just `eslint --auto-fix`

File: jest.config.js [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
-  testMatch: ["**/*.spec.ts"],
+  preset: 'ts-jest',
+  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json', 'node'],
+  testURL: 'http://localhost',
+  transformIgnorePatterns: ['[/\\\\]node_modules[/\\\\].+\\.(js|jsx)$'],
+  testMatch: ['**/*.spec.ts'],
   collectCoverage: true,
-  collectCoverageFrom: ["src/*.ts", "!./index.ts"],
-  coverageDirectory: "./coverage/",
+  collectCoverageFrom: ['src/*.ts', '!./index.ts', '!src/rules.guard.ts'],
```
Comment (gagoar) [inline jest.config.js:19]: we don't want to collect coverage from an autogenerated file.

File: package.json [hunk 3]
```diff
@@ -102,7 +102,6 @@
   "author": "Gago <xeroice@gmail.com>",
   "dependencies": {
     "@actions/core": "1.2.6",
-    "@github/memoize": "^1.0.1",
```
Comment (gagoar) [inline package.json:105]: we have the actual files instead of the library anyway.

File: package.json [hunk 4]
```diff
@@ -137,6 +136,7 @@
     "prettier-eslint": "12.0.0",
     "prettier-eslint-cli": "5.0.0",
     "pretty-quick": "3.1.0",
+    "ts-auto-guard": "1.0.0-alpha.13",
```
Comment (gagoar) [inline package.json:139]: this is what I've used to generate the type guards

**Review by cyamonide — APPROVED** [review]
looks good 🚀

**Pull request merged**